$OpenBSD: patch-src_handlers_c,v 1.4 2011/12/01 16:00:01 dcoppa Exp $

Bugfix: ignore EnterNotify events to prevent wrong focus in complex
tabbed/stacked layouts
(upstream git commit bd4b240d56df08bc255a09c9cb0d98b457c785dc)

Bugfix: skip dock clients when handling FocusIn events
(upstream git commit 5f8d719835832a91232d0331a25512e9a0853494)

--- src/handlers.c.orig	Fri Nov 11 23:41:20 2011
+++ src/handlers.c	Thu Dec  1 13:58:48 2011
@@ -452,10 +452,20 @@ static int handle_screen_change(xcb_generic_event_t *e
  *
  */
 static int handle_unmap_notify_event(xcb_unmap_notify_event_t *event) {
-    // XXX: this is commented out because in src/x.c we disable EnterNotify events
-    /* we need to ignore EnterNotify events which will be generated because a
-     * different window is visible now */
-    //add_ignore_event(event->sequence, XCB_ENTER_NOTIFY);
+    /* If the client (as opposed to i3) destroyed or unmapped a window, an
+     * EnterNotify event will follow (indistinguishable from an EnterNotify
+     * event caused by moving your mouse), causing i3 to set focus to whichever
+     * window is now visible.
+     *
+     * In a complex stacked or tabbed layout (take two v-split containers in a
+     * tabbed container), when the bottom window in tab2 is closed, the bottom
+     * window of tab1 is visible instead. X11 will thus send an EnterNotify
+     * event for the bottom window of tab1, while the focus should be set to
+     * the remaining window of tab2.
+     *
+     * Therefore, we ignore all EnterNotify events which have the same sequence
+     * as an UnmapNotify event. */
+    add_ignore_event(event->sequence, XCB_ENTER_NOTIFY);
 
     DLOG("UnmapNotify for 0x%08x (received from 0x%08x), serial %d\n", event->window, event->event, event->sequence);
     Con *con = con_by_window_id(event->window);
@@ -965,6 +975,12 @@ static int handle_focus_in(xcb_focus_in_event_t *event
 
     if (focused_id == event->event) {
         DLOG("focus matches the currently focused window, not doing anything\n");
+        return 1;
+    }
+
+    /* Skip dock clients, they cannot get the i3 focus. */
+    if (con->parent->type == CT_DOCKAREA) {
+        DLOG("This is a dock client, not focusing.\n");
         return 1;
     }
 
